tests/testrom.c

```diff
@@ -167,80 +167,85 @@ static const struct Testsuite
+	int successes = 0;
 
 	const int NUM_TESTS = sizeof(tests) / sizeof(tests[0]);
-    uint64_t start = timer_ticks();
-    for (int i=0; i < NUM_TESTS; i++) {
-    	static char logbuf[16384];
-
-    	// Prepare the test context
-    	TestContext ctx;
-    	ctx.log = logbuf;
-    	ctx.logleft = sizeof(logbuf);
-    	ctx.result = TEST_SUCCESS;
-    	rand_state = 1; // reset to be fully reproducible
-
-    	printf("%-30s", tests[i].name);
-    	fflush(stdout);
-
-    	uint64_t test_start = timer_ticks();
-
-    	// Run the test!
-    	tests[i].fn(&ctx);
-
-    	// Compute the test duration
-    	uint64_t test_stop = timer_ticks();
-    	uint64_t test_duration = (test_stop - test_start) / 1024;
-    	int64_t test_diff = (test_duration - tests[i].duration);
-    	if (test_diff < 0) test_diff = -test_diff;
-
-    	if (ctx.result == TEST_FAILED) {
-    		failures++;
-    		printf("FAIL\n\n");
-
-    		if (ctx.log != logbuf) {			
-	    		printf("%s\n\n", logbuf);
-    		}
-    	}
-    #if BENCHMARK_TESTS
-    	// If there's more than a 1% drift on the running time (/1024) compared to
-    	// the expected one, make the test fail. Something happened and we
-    	// need to double check this.
-    	else if ((float)test_diff / (float)test_duration > 0.01)
-    	{
-    		failures++;
-    		printf("FAIL\n\n");
-
-    		printf("Duration changed by %.1f%%\n", (float)test_diff * 100.0 / (float)test_duration);
-    		printf("(expected: %ldK, measured: %lldK)\n\n", tests[i].duration, test_duration);
-    	}
-    #endif
-    	else {
-    		successes++;
-    		printf("PASS\n");
-    	}
-    }
-    uint64_t stop = timer_ticks();
-
-    int64_t total_time = TIMER_MICROS(stop-start) / 1000000;
-
-    printf("\nTestsuite finished in %02lld:%02lld\n", total_time%60, total_time/60);
-    printf("Passed: %d out of %d\n", successes, NUM_TESTS);
+	uint32_t start;
+	READ_TICKS(start);
+	for (int i=0; i < NUM_TESTS; i++) {
+		static char logbuf[16384];
+
+		// Prepare the test context
+		TestContext ctx;
+		ctx.log = logbuf;
+		ctx.logleft = sizeof(logbuf);
+		ctx.result = TEST_SUCCESS;
+		rand_state = 1; // reset to be fully reproducible
+
+		printf("%-30s", tests[i].name);
+		fflush(stdout);
+
+		uint32_t test_start;
+		READ_TICKS(test_start);
+
+		// Run the test!
+		tests[i].fn(&ctx);
+
+		// Compute the test duration
+		uint32_t test_stop;
+		READ_TICKS(test_stop);
+
+		int32_t test_duration = (test_stop - test_start) / 1024;
+		int64_t test_diff = (test_duration - tests[i].duration);
+		if (test_diff < 0) test_diff = -test_diff;
+
+		if (ctx.result == TEST_FAILED) {
+			failures++;
+			printf("FAIL\n\n");
+
+			if (ctx.log != logbuf) {			
+				printf("%s\n\n", logbuf);
+			}
+		}
+	#if BENCHMARK_TESTS
+		// If there's more than a 10% drift on the running time (/1024) compared to
+		// the expected one, make the test fail. Something happened and we
+		// need to double check this.
+		else if (tests[i].duration > 0 && ((float)test_diff / (float)test_duration > 0.1))
```

Comment:
10% sounds like a lot to keep as a performance test. Can you elaborate on why this is needed?

Reply:
Because on my tests on an everdrive they tend to be off by 7-9%. Couldnt find a reason for that. I think having more slack for IO benchmarks is the only way here.

Reply:
I see, it is probably the case that I/O speed with different SDKs is really unpredictable (eg: the SD card might be slower or faster). I would suggest to add a flag to mark those tests as "I/O TESTS" in the tests structure (this flag set can also be useful in the future), and use that flag to raise the allowed drift to 10% for now.

I now we have a too little testsuite for now, but my hope is that for many tests we can keep some kind of semi-perfect timing. Maybe I'm optimistic and we will have to change all of them to 10%, but I'd try to keep this assumption for now.